apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/WriterThreadPool.java

```diff
@@ -0,0 +1,237 @@
+package gov.cms.bfd.pipeline.rda.grpc.sink.concurrent;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.hash.HashFunction;
+import com.google.common.hash.Hashing;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import gov.cms.bfd.pipeline.rda.grpc.ProcessingException;
+import gov.cms.bfd.pipeline.rda.grpc.RdaSink;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import javax.annotation.Nonnull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Manages a pool of worker threads to write claim and sequence number updates to a database
+ * concurrently. The lifecycle of the pool and all of its resources are tied to the lifecycle of
+ * this object so calling close() method clears up all resources.
+ *
+ * @param <TMessage> RDA API message class
+ * @param <TClaim> JPA claim entity class
+ */
+public class WriterThreadPool<TMessage, TClaim> implements AutoCloseable {
+  private static final Logger LOGGER = LoggerFactory.getLogger(WriterThreadPool.class);
+  private static final HashFunction Hasher = Hashing.goodFastHash(32);
+
+  private final SequenceNumberTracker sequenceNumbers;
+  private final RdaSink<TMessage, TClaim> sink;
+  private final ExecutorService threadPool;
+  private final BlockingQueue<ReportingCallback.ProcessedBatch<TMessage>> outputQueue;
+  private final SequenceNumberWriterThread<TMessage, TClaim> sequenceNumberWriter;
+  private final List<ClaimWriterThread<TMessage, TClaim>> writers;
+
+  public WriterThreadPool(
+      int maxThreads, int batchSize, Supplier<RdaSink<TMessage, TClaim>> sinkFactory) {
+    Preconditions.checkArgument(maxThreads > 0);
+    Preconditions.checkArgument(batchSize > 0);
+    sequenceNumbers = new SequenceNumberTracker(0);
+    sink = sinkFactory.get();
+    threadPool =
+        Executors.newCachedThreadPool(
```

Comment:
Can you elaborate some on your choice of a cached thread pool here as opposed to a fixed thread pool? Have we been able to load test this at production levels and ensure that the size of this pool was kept to reasonable levels? It might be useful to include somewhere in here as a comment as well the factors that led to choosing this.

Reply:
Cached thread pool creates only as many threads as needed to execute submitted jobs and shuts down any unused/cached threads if they are idle very long.  I prefer them here because it's easy to get the count wrong when creating a fixed thread pool and have a stuck job that never launches or an unused thread that's sitting idle and wasting resources.  The count issue is particularly likely when a class is updated later to add a new type worker thread.

In this case the pool is created and all jobs are submitted at the beginning so there is no way for the pool to run away on us.  It will always have exactly one thread per job.  Which in this case is one thread per ClaimWriter or SequenceNumberWriter.